File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        li1 = []
        li2 = []
        
        while headA:
            li1.append(headA)
            headA = headA.next
        
        while headB:
            li2.append(headB)
            headB = headB.next
        
        prev = None
        while li1 and li2:
            nodesa = li1.pop()
            nodesb = li2.pop()

            if nodesa != nodesb:
                return prev
            prev = nodesa
        
        return prev
        
        
        

File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        vector<ListNode*> v1, v2;

        while (headA != nullptr) {
            v1.push_back(headA);
            headA = headA->next;
        }

        while (headB != nullptr) {
            v2.push_back(headB);
            headB = headB->next;
        }

        ListNode* prev = nullptr;
        while (!v1.empty() && !v2.empty()) {
            ListNode* nodeA = v1.back();
            ListNode* nodeB = v2.back();
            v1.pop_back();
            v2.pop_back();

            if (nodeA != nodeB) {
                return prev;
            }
            prev = nodeA;
        }

        return prev;
    }
};